nodes/basic-pow/src/service.rs

```diff
@@ -0,0 +1,176 @@
+pub use sc_executor::NativeExecutor;
+use crate::pow::Sha3Algorithm;
+use sc_network::{config::DummyFinalityProofRequestBuilder};
+
+// Our native executor instance.
+native_executor_instance!(
+	pub Executor,
+	runtime::api::dispatch,
+	runtime::native_version,
+);
+
+pub fn build_inherent_data_providers() -> Result<InherentDataProviders, ServiceError> {
+	let providers = InherentDataProviders::new();
+
+	providers
+		.register_provider(sp_timestamp::InherentDataProvider)
+		.map_err(Into::into)
+		.map_err(sp_consensus::error::Error::InherentData)?;
+
+	Ok(providers)
+}
+
+/// Starts a `ServiceBuilder` for a full service.
+///
+/// Use this macro if you don't actually need the full service, but just the builder in order to
+/// be able to perform chain operations.
+macro_rules! new_full_start {
+	($config:expr) => {{
+		let mut import_setup = None;
+		let inherent_data_providers = crate::service::build_inherent_data_providers()?;
+
+		let builder = sc_service::ServiceBuilder::new_full::<
+			runtime::opaque::Block, runtime::RuntimeApi, crate::service::Executor
+		>($config)?
+			.with_select_chain(|_config, backend| {
+				Ok(sc_client::LongestChain::new(backend.clone()))
+			})?
+			.with_transaction_pool(|config, client, _fetcher| {
+				let pool_api = sc_transaction_pool::FullChainApi::new(client.clone());
+				Ok(sc_transaction_pool::BasicPool::new(config, std::sync::Arc::new(pool_api)))
+			})?
+			.with_import_queue(|_config, client, select_chain, _transaction_pool| {
+
+				let pow_block_import = sc_consensus_pow::PowBlockImport::new(
+					client.clone(),
+					client.clone(),
+					crate::pow::Sha3Algorithm,
+					0, // check inherents starting at block 0
+					select_chain,
+					inherent_data_providers.clone(),
+				);
+
+				let import_queue = sc_consensus_pow::import_queue(
+					Box::new(pow_block_import.clone()),
+					crate::pow::Sha3Algorithm,
+					inherent_data_providers.clone(),
+				)?;
+
+				import_setup = Some(pow_block_import);
+
+				Ok(import_queue)
+			})?;
+
+		(builder, import_setup, inherent_data_providers)
+	}}
+}
+
+/// Builds a new service for a full client.
+pub fn new_full(config: Configuration<GenesisConfig>)
+	-> Result<impl AbstractService, ServiceError>
+{
+	let is_authority = config.roles.is_authority();
+
+	// sentry nodes announce themselves as authorities to the network
+	// and should run the same protocols authorities do, but it should
+	// never actively participate in any consensus process.
+	let participates_in_consensus = is_authority && !config.sentry_mode;
+
+	let (builder, mut import_setup, inherent_data_providers) = new_full_start!(config);
+	let block_import = import_setup.take()
+		.expect("Block Import is present for Full Services or setup failed before. qed");
+
+	let service = builder
+		.with_finality_proof_provider(|_client, _backend|
+			// Question. Why bother giving () as a finality proof provider when I can just
+			// not call `with_finality_proof_provider` at all?
```

Comment:
I believe you can (just not call it). Previously it was there because omitting it would result in Rust not able to resolve some generic parameters.